engine/kv_engine.cpp

```diff
@@ -1720,3 +1736,193 @@ Status KVEngine::RestoreDlistRecords(DLRecord *pmp_record) {
+  }
+  bool pop_success = false;
+  switch (where) {
+  case QueueOpWhere::Left: {
+    pop_success = queue_ptr->PopFront(value);
+    break;
+  }
+  case QueueOpWhere::Right: {
+    pop_success = queue_ptr->PopBack(value);
+    break;
+  }
+  default: {
+    GlobalLogger.Error("Impossible!");
+    return Status::Abort;
+  }
+  }
+  if (pop_success) {
+    return Status::Ok;
+  } else {
+    return Status::NotFound;
+  }
+}
+
+Status KVEngine::xPush(StringView const collection_name, StringView const value,
+                       KVEngine::QueueOpWhere where) try {
+  Status s = MaybeInitWriteThread();
+  if (s != Status::Ok) {
+    return s;
+  }
+
+  Queue *queue_ptr = nullptr;
+
+  // Find UnorederedCollection, create if none exists
+  {
+    queue_ptr = findQueue(collection_name);
+    if (!queue_ptr) {
+      HashTable::KeyHashHint hint_collection =
+          hash_table_->GetHint(collection_name);
+      std::unique_lock<SpinMutex> lock_collection{*hint_collection.spin};
+      {
+        // Lock and find again in case other threads have created the
+        // UnorderedCollection
+        queue_ptr = findQueue(collection_name);
+        if (queue_ptr) {
+          // Some thread already created the collection
+          // Do nothing
+        } else {
+          auto queue_uptr = createQueue(collection_name);
+          queue_ptr = queue_uptr.get();
+          {
+            std::lock_guard<std::mutex> lg{list_mu_};
+            queue_uptr_vec_.emplace_back(queue_uptr.release());
+          }
+
+          HashEntry hash_entry_collection;
+          HashEntry *p_hash_entry_collection = nullptr;
+          Status s = hash_table_->SearchForWrite(
+              hint_collection, collection_name, RecordType::QueueRecord,
+              &p_hash_entry_collection, &hash_entry_collection, nullptr);
+          kvdk_assert(s == Status::NotFound, "Logically impossible!");
+          hash_table_->Insert(
+              hint_collection, p_hash_entry_collection, RecordType::QueueRecord,
+              reinterpret_cast<uint64_t>(queue_ptr), HashOffsetType::Queue);
+        }
+      }
+    }
+  }
+
+  // Push
+  {
+    TimeStampType ts = get_timestamp();
+    switch (where) {
+    case QueueOpWhere::Left: {
+      queue_ptr->PushFront(ts, value);
+      return Status::Ok;
+    }
+    case QueueOpWhere::Right: {
+      queue_ptr->PushBack(ts, value);
+      return Status::Ok;
+    }
+    default: {
+      GlobalLogger.Error("Impossible!");
+      return Status::Abort;
+    }
+    }
+  }
+} catch (std::bad_alloc const &ex) {
+  return Status::PmemOverflow;
+}
+
+Status KVEngine::RestoreQueueRecords(DLRecord *pmp_record) {
+  switch (pmp_record->entry.meta.type) {
+  case RecordType::QueueRecord: {
+    Queue *queue_ptr = nullptr;
+    std::lock_guard<std::mutex> lg{list_mu_};
+    {
+      queue_uptr_vec_.emplace_back(
+          std::make_unique<Queue>(pmem_allocator_.get(), pmp_record));
+      queue_ptr = queue_uptr_vec_.back().get();
```

Comment:
I don‘t understand: you use `unique_ptr`, why not directly use `unique_ptr->Name()`, instead of call `get()` to transfer to raw pointer?

Reply:
Queue* is needed for updating HashTable.

Reply:
all right, It's a little redundant. Only pass raw pointer when call 
`hash_table_->Insert(
              hint_collection, p_hash_entry_collection, RecordType::QueueRecord,
              reinterpret_cast<uint64_t>(unique_ptr.get()), HashOffsetType::Queue);`